Apps/g.Tec/g.NEEDaccess/src/mainwindow.cpp

```diff
@@ -0,0 +1,557 @@
+	connect(m_pTimer, SIGNAL(timeout()), this, SLOT(notify_samples_pushed()));
+}
+
+
+MainWindow::~MainWindow()
+{
+	if (m_bStreaming)
+	{
+		handleResult("GDS_StopAcquisition", GDS_StopAcquisition(m_connectionHandle));
+		handleResult("GDS_StopStreaming", GDS_StopStreaming(m_connectionHandle));
+		handleResult("GDS_SetDataReadyCallback",
+			GDS_SetDataReadyCallback(m_connectionHandle, NULL, 0, NULL));
+		delete[] m_dataBuffer;
+		m_dataBuffer = NULL;
+		delete this->m_eegOutlet;
+		this->m_eegOutlet = NULL;
+	}
+	if (m_bConnected)
+	{
+		handleResult("GDS_Disconnect",
+			GDS_Disconnect(&m_connectionHandle));
+	}
+
+	GDS_Uninitialize();
+    delete ui;
+	delete m_pTimer;
+	m_pTimer = nullptr;
+}
+
+
+void MainWindow::on_scanPushButton_clicked()
+{
+	bool success = true;
+
+    ui->scanPushButton->setText("Scanning...");
+    ui->scanPushButton->setDisabled(true);
+
+	strcpy(this->m_hostEndpoint.IpAddress, ui->lineEdit_serverip->text().toLatin1().data());
+	this->m_hostEndpoint.Port = ui->serverPortSpinBox->value();
+	this->m_localEndpoint.Port = ui->clientPortSpinBox->value();	
+
+	GDS_DEVICE_CONNECTION_INFO* connected_devices = NULL;
+	size_t count_daq_units = 0;
+	handleResult("GDS_GetConnectedDevices",
+		GDS_GetConnectedDevices(m_hostEndpoint, m_localEndpoint, &connected_devices, &count_daq_units));
+
+	std::vector<GDS_DeviceInfo> new_devices;
+	uint32_t count_devices = 0;
+	for (size_t i = 0; i < count_daq_units; i++)
+	{
+		// Each daq unit operates a different number of devices. The second "for" loop 
+		// addresses the devices attached to the daq unit.
+		for (size_t j = 0; j < connected_devices[i].ConnectedDevicesLength; j++)
+		{
+			GDS_DeviceInfo new_info;
+			new_info.index = count_devices;
+			new_info.inUse = bool(connected_devices[i].InUse) == true;
+			new_info.name = connected_devices[i].ConnectedDevices[j].Name;
+			new_info.type = connected_devices[i].ConnectedDevices[j].DeviceType;
+			new_devices.push_back(new_info);
+			count_devices++;
+		}
+	}
+	success &= handleResult("GDS_FreeConnectedDevicesList",
+		GDS_FreeConnectedDevicesList(&connected_devices, count_daq_units));
+
+	if (new_devices.size() > 0)
+	{
+		ui->availableListWidget->clear();
+		QStringList newDeviceNames;
+		for (auto it = new_devices.begin(); it != new_devices.end(); ++it) {
+			ui->availableListWidget->addItem(it->name);
+			// TODO: Use other GDS_DeviceInfo fields
+		}
+		ui->availableListWidget->setCurrentRow(0);
+	}
+	ui->scanPushButton->setText("Scan");
+	ui->scanPushButton->setDisabled(false);
+}
+
+
+void MainWindow::on_availableListWidget_itemSelectionChanged()
+{
+	ui->connectPushButton->setDisabled(false);
+}
+
+
+void MainWindow::on_connectPushButton_clicked()
+{
+	ui->goPushButton->setDisabled(true);
+	QString pbtext = ui->connectPushButton->text();
+	if (pbtext == "Connect")
+	{
+		QString deviceName = ui->availableListWidget->currentItem()->text();
+		char(*device_names)[DEVICE_NAME_LENGTH_MAX] = new char[1][DEVICE_NAME_LENGTH_MAX];
```

Comment:
I think a using a `std::string` would be safer:
<pre><code>std::string deviceNameStr(deviceName.toLatin1().toStdString());
char* deviceNames[1] = {deviceNameStr.c_str()};</code></pre>

Reply:
Now I remember why I did it this crazy way: That variable is passed to a function with its argument declaration `char (*deviceNames)[DEVICE_NAME_LENGTH_MAX]`. I can't figure out how to get that type from std::string, nor do I want to because I think it's an incorrect design in the API. If you notice, that's actually DEVICE_NAME_LENGTH_MAX (=32) different char arrays, each with N characters (here N == 1). When you view this variable in the debugger inspector, it doesn't show characters running across a line, rather it has characters running down a column.

I'll add some comments to this section so no one else bangs their head against the wall as I did.

Reply:
But in this case it's a one element array, so the heap allocation is not strictly necessary.
A stack var could also work: <pre><code>char deviceNameCStr[DEVICE_NAME_LENGTH_MAX];
strncpy(deviceNameCStr, deviceName, DEVICE_NAME_LENGTH_MAX);
//[..]
SomeApiFunc(..., &deviceNameCStr); //pointer to device name string</code></pre>
Maybe the [Cpp Core Guidelines Library](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f25-use-a-zstring-or-a-not_nullzstring-to-designate-a-c-style-string) could error proof this, but it's an additional dependency.

I'd also hope the API uses null-terminated strings and doesn't blindly copy `DEVICE_NAME_LENGTH_MAX` elements, but I wouldn't bet on it.

Reply:
Ok, disregard that. I tried it with a sample program and it's really as weird as you describe it.